File: rust/src/huffman.rs
use crate::lz;

const MAX_TREE_HT: usize = 100;

#[derive(Debug, Clone)]
pub struct MinHeapNodeLZ77 {
    pub tuple: lz::LZ77Tuple,    // Mesma estrutura LZ77Tuple
    pub freq: usize,           // Correspondente a unsigned em C
    pub left: Option<Box<MinHeapNodeLZ77>>,  // Ponteiro opcional para outro MinHeapNodeLZ77
    pub right: Option<Box<MinHeapNodeLZ77>>, // Ponteiro opcional para outro MinHeapNodeLZ77
}


#[derive(Debug, Clone)]
pub struct MinHeap {
    pub size: usize,                     // Correspondente a unsigned em C
    pub capacity: usize,                 // Correspondente a unsigned em C
    pub array: Vec<Box<MinHeapNodeLZ77>>, // Vetor de ponteiros para MinHeapNodeLZ77
}

fn new_node_lz77(tuple: lz::LZ77Tuple, freq: usize) -> Box<MinHeapNodeLZ77> {
    Box::new(MinHeapNodeLZ77 {
        tuple,
        freq,
        left: None,
        right: None,
    })
}

fn create_min_heap(capacity: usize) -> MinHeap {
    MinHeap {
        size: 0,
        capacity,
        array: Vec::with_capacity(capacity),
    }
}


fn min_heapify(min_heap: &mut MinHeap, idx: usize) {
    let mut smallest = idx;
    let left = 2 * idx + 1;
    let right = 2 * idx + 2;

    if left < min_heap.size && min_heap.array[left].freq < min_heap.array[smallest].freq {
        smallest = left;
    }

    if right < min_heap.size && min_heap.array[right].freq < min_heap.array[smallest].freq {
        smallest = right;
    }

    if smallest != idx {
        min_heap.array.swap(smallest, idx);
        min_heapify(min_heap, smallest);
    }
}

fn extract_min(min_heap: &mut MinHeap) -> Box<MinHeapNodeLZ77> {
    let temp = min_heap.array.swap_remove(0);
    min_heap.size -= 1;
    min_heapify(min_heap, 0);
    temp
}

fn insert_min_heap(min_heap: &mut MinHeap, min_heap_node: Box<MinHeapNodeLZ77>) {
    min_heap.size += 1;
    let mut i = min_heap.size - 1;
    while i > 0 && min_heap_node.freq < min_heap.array[(i - 1) / 2].freq {
        if i < min_heap.array.len() {
            min_heap.array[i] = min_heap.array[(i - 1) / 2].clone();
        } else {
            min_heap.array.push(min_heap.array[(i - 1) / 2].clone());
        }
        i = (i - 1) / 2;
    }
    if i < min_heap.array.len() {
        min_heap.array[i] = min_heap_node;
    } else {
        min_heap.array.push(min_heap_node);
    }
}

fn build_min_heap(min_heap: &mut MinHeap) {
    let n = min_heap.size - 1;
    for i in (0..=((n - 1) / 2)).rev() {
        min_heapify(min_heap, i);
    }
}

fn build_huffman_tree_lz77(data: &[lz::LZ77Tuple], freq: &[usize]) -> Box<MinHeapNodeLZ77> {
    let mut min_heap = create_min_heap(data.len());
    for i in 0..data.len() {
        min_heap.array.push(new_node_lz77(data[i], freq[i]));
    }
    min_heap.size = data.len();
    build_min_heap(&mut min_heap);

    while min_heap.size > 1 {
        let left = extract_min(&mut min_heap);
        let right = extract_min(&mut min_heap);

        let combined_tuple = lz::LZ77Tuple { offset: 0, length: 0, next_char: '$' };
        let mut top = new_node_lz77(combined_tuple, left.freq + right.freq);
        top.left = Some(left);
        top.right = Some(right);

        insert_min_heap(&mut min_heap, top);
    }

    extract_min(&mut min_heap)
}

pub fn compress_with_huffman(data: &[lz::LZ77Tuple], freq: &[usize], codes: &mut Vec<String>) {
    let root = build_huffman_tree_lz77(data, freq);

    let mut arr = Vec::with_capacity(MAX_TREE_HT);
    let  top = 0;

    generate_huffman_codes(&root, &mut arr, top, codes);

    for code in codes.iter() {
        print!("{}", code);
    }
    println!();
}


fn generate_huffman_codes(
    node: &Box<MinHeapNodeLZ77>,
    arr: &mut Vec<i32>,
    top: usize,
    codes: &mut Vec<String>,
) {
    if let Some(ref left) = node.left {
        arr.push(0);
        generate_huffman_codes(left, arr, top + 1, codes);
        arr.pop();
    }

    if let Some(ref right) = node.right {
        arr.push(1);
        generate_huffman_codes(right, arr, top + 1, codes);
        arr.pop();
    }

    if node.left.is_none() && node.right.is_none() {
        let code = arr.iter().map(|&bit| (bit + b'0' as i32) as u8 as char).collect::<String>();
        codes.push(code);
    }
}

File: C/huffman.c

#include "process.h"
// Estrutura para armazenar as tuplas do LZ77

// Função para criar um novo nó de árvore de Huffman para LZ77
struct MinHeapNodeLZ77* newNodeLZ77(LZ77Tuple tuple, unsigned freq) {
    struct MinHeapNodeLZ77* temp = (struct MinHeapNodeLZ77*)malloc(sizeof(struct MinHeapNodeLZ77));
    temp->left = temp->right = NULL;
    temp->tuple = tuple;
    temp->freq = freq;
    return temp;
}


// Função para criar uma heap mínima
struct MinHeap* createMinHeap(unsigned capacity) {
    struct MinHeap* minHeap = (struct MinHeap*)malloc(sizeof(struct MinHeap));
    minHeap->size = 0;
    minHeap->capacity = capacity;
    minHeap->array = (struct MinHeapNodeLZ77**)malloc(minHeap->capacity * sizeof(struct MinHeapNodeLZ77*));
    return minHeap;
}

// Função para trocar dois nós
void swapMinHeapNode(struct MinHeapNodeLZ77** a, struct MinHeapNodeLZ77** b) {
    struct MinHeapNodeLZ77* t = *a;
    *a = *b;
    *b = t;
}

// Função para garantir que a propriedade da heap mínima seja mantida
void minHeapify(struct MinHeap* minHeap, int idx) {
    int smallest = idx;
    int left = 2 * idx + 1;
    int right = 2 * idx + 2;

    if (left < minHeap->size && minHeap->array[left]->freq < minHeap->array[smallest]->freq)
        smallest = left;

    if (right < minHeap->size && minHeap->array[right]->freq < minHeap->array[smallest]->freq)
        smallest = right;

    if (smallest != idx) {
        swapMinHeapNode(&minHeap->array[smallest], &minHeap->array[idx]);
        minHeapify(minHeap, smallest);
    }
}

// Função para extrair o nó de menor frequência da heap mínima
struct MinHeapNodeLZ77* extractMin(struct MinHeap* minHeap) {
    struct MinHeapNodeLZ77* temp = minHeap->array[0];
    minHeap->array[0] = minHeap->array[minHeap->size - 1];
    --minHeap->size;
    minHeapify(minHeap, 0);
    return temp;
}

// Função para inserir um nó na heap mínima
void insertMinHeap(struct MinHeap* minHeap, struct MinHeapNodeLZ77* minHeapNode) {
    ++minHeap->size;
    int i = minHeap->size - 1;
    while (i && minHeapNode->freq < minHeap->array[(i - 1) / 2]->freq) {
        minHeap->array[i] = minHeap->array[(i - 1) / 2];
        i = (i - 1) / 2;
    }
    minHeap->array[i] = minHeapNode;
}

// Função para construir a heap mínima a partir de um array de nós de árvore
void buildMinHeap(struct MinHeap* minHeap) {
    int n = minHeap->size - 1;
    int i;
    for (i = (n - 1) / 2; i >= 0; --i)
        minHeapify(minHeap, i);
}

// Função para construir a árvore de Huffman a partir das tuplas do LZ77
struct MinHeapNodeLZ77* buildHuffmanTreeLZ77(LZ77Tuple data[], int freq[], int size) {
    struct MinHeapNodeLZ77 *left, *right, *top;

    // Criação e construção da heap mínima (MinHeap)
    struct MinHeap* minHeap = createMinHeap(size);
    for (int i = 0; i < size; ++i)
        minHeap->array[i] = newNodeLZ77(data[i], freq[i]);
    minHeap->size = size;
    buildMinHeap(minHeap);

    // Construção da árvore de Huffman
    while (minHeap->size != 1) {
        left = extractMin(minHeap);
        right = extractMin(minHeap);

        LZ77Tuple combinedTuple = {0, 0, '$'};  // Símbolo placeholder
        top = newNodeLZ77(combinedTuple, left->freq + right->freq);
        top->left = left;
        top->right = right;

        insertMinHeap(minHeap, top);
    }

    // O nó restante na heap é a raiz da árvore de Huffman
    return extractMin(minHeap);
}

// Função para gerar os códigos Huffman a partir da árvore de Huffman
void generateHuffmanCodes(struct MinHeapNodeLZ77* root, int arr[], int top, char** codes, int* index) {
    if (root->left) {
        arr[top] = 0;
        generateHuffmanCodes(root->left, arr, top + 1, codes, index);
    }

    if (root->right) {
        arr[top] = 1;
        generateHuffmanCodes(root->right, arr, top + 1, codes, index);
    }

    if (!(root->left) && !(root->right)) {
        codes[*index] = (char*)malloc(top + 1);
        for (int i = 0; i < top; ++i) {
            codes[*index][i] = arr[i] + '0';
        }
        codes[*index][top] = '\0';
        (*index)++;
    }
}

// Função para codificar o texto original usando os códigos Huffman
void compressWithHuffman(LZ77Tuple data[], int freq[], int size, char** codes, int tuple_count) {
    struct MinHeapNodeLZ77* root = buildHuffmanTreeLZ77(data, freq, size);

    int arr[MAX_TREE_HT], top = 0;
    int index = 0;
    generateHuffmanCodes(root, arr, top, codes, &index);
    // for (int i = 0; i < tuple_count; i++) {
    //     // printf("%s", codes[i]);
    // }
    // printf("\n");
}